dbus/xdg.c

```diff
@@ -217,6 +217,79 @@ static int handle_notify(sd_bus_message *msg, void *data,
 				return ret;
 			}
 			notif->progress = progress;
+		} else if (strcmp(hint, "image-data") == 0) {
+			ret = sd_bus_message_enter_container(msg, 'v', "(iiibiiay)");
+			if (ret < 0) {
+				return ret;
+			}
+
+			ret = sd_bus_message_enter_container(msg, 'r', "iiibiiay");
+			if (ret < 0) {
+				return ret;
+			}
+
+			struct iiibiiay *image_data = calloc(1, sizeof(struct iiibiiay));
+			if (image_data == NULL) {
+				return -1;
+			}
+
+			ret = sd_bus_message_read(msg, "iiibii", &image_data->width,
+					&image_data->height, &image_data->rowstride,
+					&image_data->has_alpha, &image_data->bits_per_sample,
+					&image_data->channels);
+			if (ret < 0) {
+				free(image_data);
+				return ret;
+			}
+
+			size_t image_len = (image_data->height - 1) * image_data->rowstride +
+				image_data->width * ((image_data->channels *
+				image_data->bits_per_sample + 7) / 8);
```

Comment:
Hmm. I'm not sure I get this formula.

Reply:
I've added some comments in code to explain that.

Reply:
Why is the last row special?

Reply:
> If you are doing memcpy() of raw pixbuf data, note that the last row in the pixbuf may not be as wide as the full rowstride, but rather just as wide as the pixel data needs to be. That is, it is unsafe to do memcpy (dest, pixels, rowstride * height) to copy a whole pixbuf. Use gdk_pixbuf_copy() instead, or compute the width in bytes of the last row as width * ((n_channels * bits_per_sample + 7) / 8).
(Quote from GdkPixbuf)

GdkPixBuf may trim the last row to save space.
Libnotify and notification-daemon also use this formula.
Not sure if there is any client implementation that send untrimmed data, maybe we should just ignore extra bits instead of an error?

Reply:
>GdkPixBuf may trim the last row to save space.

I don't understand how they can do so. If the image contains 1000 pixels, you can't "trim the last row" and only send 998 pixels.

>maybe we should just ignore extra bits instead of an error?

Or do you mean they pad the image data (adding extra bytes we shouldn't care about)?

In this case we should just ignore extra bytes, yes.

Reply:
Yes I meant extra padding bytes may exist in the last row.

Reply:
I've updated the commit to ignore extra bytes, also changed the comment style to fit with other comments.

Reply:
Then this formula can be simplified to `image_data->height * image_data->rowstride`?

Reply:
`image_data->rowstride` contains the extra padding bytes, so`image_data->height * image_data->rowstride` contains the extra bytes in the last row. Client may not send extra bytes in last row, so we should expect the length to be without that.

Reply:
In other words: the last row may not be padded to `rowstride`?

Gah, this sucks. Okay, let's just keep this version.